win/TUI: Improve terminal/console support

justinmk opened the pull request:
- Merge #9094 **except** 367a60a7b3a0 and e931f72cdd49 .
- Depends on `uv_set_vterm_state()` which we are trying to upstream. https://github.com/libuv/libuv/pull/1873  https://github.com/libuv/libuv/pull/1884
- Adds builtin terminfos for ConEmu, cygwin, and cmd.exe

erw7 committed:
win/TUI: SetConsoleMode() to override libuv #9094

Use uv_set_vterm_state() to override libuv's guess.
See https://github.com/libuv/libuv/pull/1873/ for discussion.

This commit uses a terminal-detection approach based on
GetProcessImageFileNameW(...), which will be reverted in the following
commit. The approach was intended to handle the case of running in
winpty (:terminal), but we will add $NVIM env var for that.

Also add some support for ConEmu, cygwin.

erw7 committed:
win/TUI: ConEmu: set immediate_wrap_after_last_column #9094

erw7 committed:
win/TUI: builtin terminfos for cygwin, conemu, et al.

erw7 committed:
win/TUI: Fix scroll on Windows legacy console

justinmk committed:
Merge 'upstream/master' into pr-win-erw7

justinmk committed:
test: win/TUI builtin terminfos

justinmk committed:
update_terminfo.sh: Use printf instead of echo

justinmk committed:
update_terminfo.sh: NOLINT data arrays

jamessan commented:
> When I run the script I see these messages, have you seen this before?

I have.  It's nothing to worry about.

jamessan commented on scripts/update_terminfo.sh:
```diff
@@ -72,18 +76,19 @@ EOF
   path="$(find "$db" -name "$term")"
-  if [[ -z $path ]]; then
-    echo "Not found: $term. Skipping." 1>&2
+  if [ -z $path ]; then
```
If you're removing the `[[` bashism (even though it's a bash script), then you need to quote `$path`.

```suggestion
  if [ -z "$path" ]; then
```

mhinz reviewed (commented):
The printf commit seemst like code churn, since it really just replaces one command with another without real gain.

I approve of the NOLINT change, though. :)

mhinz commented:
> > since it really just replaces one command with another without real gain.
> 
> `echo -n` is non-standard. `printf` is the appropriate way to handle that and also handles string interpolation better.

But `-n` is part of the Bash `echo` builtin.

justinmk commented:
> Too bad it didn't work. :( Not sure where the linter expects the comment to be.

It worked, but the remaining lint errors are regarding the docstrings above each of the data lines, for example:

    //  acs_chars=+^P\054^Q-^X.^Y0\333`^Da\261f\370g\361h\260j\331k\277l\332m\300n\305o~p\304q\304r\304s_t\303u\264v\301w\302x\263y\363z\362{\343|\330}\234~\376,

> But -n is part of the Bash echo builtin.

printf is "best practice" and works the same everywhere. It's good to have a proliferation of best-practice examples in a codebase, because existing code tends to be copied for new code.

justinmk committed:
refactor: Extract os_tty_guess_term()

- Also remove feature-detection of uv_set_vterm_state(): instead, on
  Windows we always require libuv to have that function.

justinmk committed:
cleanup: Remove os_term_is_nice()

justinmk commented:
@erw7 For reference: see 8ffd5ab6414136a7d8a093d5a8c1fd1acf221453

erw7 commented on src/nvim/os/tty.c:
```diff
@@ -0,0 +1,61 @@
+#endif
+
+#ifdef WIN32
+# if defined(WIN32) && !defined(ENABLE_VIRTUAL_TERMINAL_PROCESSING)
```
`#defined(WIN32)` is redundant

erw7 commented on src/nvim/os/tty.c:
```diff
@@ -0,0 +1,61 @@
+    // Force TERM=win32con when running in winpty.
+    *term = "win32con";
+  } else {
+    conemu_ansi = strequal(os_getenv("ConEmuANSI"), "ON");
```
This call is redundant.

erw7 commented on src/nvim/os/tty.c:
```diff
@@ -0,0 +1,61 @@
+    }
+  }
+
+  if (*term == NULL) {
```
Can we move to the else clause above?

justinmk commented on src/nvim/os/tty.c:
```diff
@@ -0,0 +1,61 @@
+      *term = "win32con";
+    }
+  }
+
```
I think for WIndows, we should assume/require it. I hope it will be upstreamed, but meanwhile anyone building on Windows will probably use our "bundled" deps.

justinmk commented:
@erw7 Thanks for the review. I pushed an update. Will merge if there are no more comments.

justinmk committed:
win/TUI: ConEmu: get back_color_erase from terminfo

We already set back_color_erase in our builtin terminfo
(terminfo_defs.h:conemu_terminfo), so we don't need to set it explicitly
in patch_terminfo_bugs().

erw7 commented on src/nvim/tui/tui.c:
```diff
@@ -1555,10 +1554,6 @@ static void patch_terminfo_bugs(TUIData *data, const char *term,
 
-  if (conemu) {
-    unibi_set_bool(ut, unibi_back_color_erase, true);
-  }
```
I think BCE will always be off in builtin terminfo(see [terminfo.c#L142](https://github.com/neovim/neovim/blob/master/src/nvim/tui/terminfo.c#L142)). ConEmu seems to have BCE so I want to enable BCE even in the case of builtin terminfo.

justinmk replied:
@erw7 ah, good catch. Let's add the exception there instead, so that the decision is in one place.

justinmk replied:
@erw7 see https://github.com/neovim/neovim/pull/9443 .
